src/capellambse_context_diagrams/_elkjs.py

```diff
@@ -299,136 +297,180 @@ class ELKOutputEdge(ELKOutputElement):
-
-
-def call_elkjs(elk_model: ELKInputData) -> ELKOutputData:
-    """Call into elk.js to auto-layout the ``diagram``.
-
-    Parameters
-    ----------
-    elk_model
-        The diagram data, sans layouting information
-
-    Returns
-    -------
-    layouted_diagram
-        The diagram data, augmented with layouting information
-    """
-    _find_node_and_npm()
-    _install_required_npm_pkg_versions()
-
-    ELKInputData.model_validate(elk_model, strict=True)
-    proc = subprocess.run(
-        ["node", str(PATH_TO_ELK_JS)],
-        executable=shutil.which("node"),
-        capture_output=True,
-        check=False,
-        input=elk_model.model_dump_json(exclude_defaults=True),
-        text=True,
-        env={**os.environ, "NODE_PATH": str(NODE_HOME)},
-    )
-    if proc.returncode or proc.stderr:
-        log.getChild("node").error("%s", proc.stderr.splitlines()[0])
-        raise NodeJSError("elk.js process failed")
 
-    return ELKOutputData.model_validate_json(proc.stdout, strict=True)
+        line = self._proc.stdout.readline()
+        if line.strip() != "--- ELK layouter started ---":
+            raise RuntimeError(
+                "Failed to start elk.js helper process using deno"
+            )
+        log.debug("Spawned elk.js helper process using deno")
+
+    def spawn_process(self):
+        """Spawn the elk.js process.
+
+        The preferenced order is:
+        binary (downloaded) > deno > binary (needs download)
+        """
+
+        if self.binary_path.exists() or shutil.which("deno") is None:
+            self._spawn_process_binary()
+        else:
+            self._spawn_process_deno()
+
+    def terminate_process(self):
+        log.debug("Terminating elk.js helper process")
+        if self._proc is not None:
+            self._proc.terminate()
+            self._proc = None
+            log.debug("Terminated elk.js helper process")
+        else:
+            log.debug("No elk.js helper process to terminate")
+
+    def get_process(self) -> subprocess.Popen:
+        if self._proc is None:
+            self.spawn_process()
+            assert self._proc is not None
+        return self._proc
+
+    def call_elkjs(self, elk_model: ELKInputData) -> ELKOutputData:
+        """Call into elk.js to auto-layout the ``diagram``.
+
+        Parameters
+        ----------
+        elk_model
+            The diagram data, sans layouting information
+
+        Returns
+        -------
+        layouted_diagram
+            The diagram data, augmented with layouting information
+        """
+        process = self.get_process()
+
+        if not process.stdin or not process.stdout:
+            raise RuntimeError("ELK process stdin/stdout not available")
+
+        ELKInputData.model_validate(elk_model, strict=True)
+        process.stdin.write(
+            elk_model.model_dump_json(exclude_defaults=True) + "\n"
+        )
+        process.stdin.flush()
+        response = process.stdout.readline()
```

Comment:
This block should be protected with a lock, to make sure only one thread is trying to communicate over this pipe at a time. We can't be sure that downstream consumers are always single-threaded, and especially for web servers (like the recently released CCD demonstrator app, or the model explorer) they often aren't.

Whether or not we want to allow multi-threading I would argue is out of scope for this discussion.